capstone/src/main/java/com/step902020/capstone/Recommender.java

```diff
@@ -0,0 +1,73 @@
+
+package com.step902020.capstone;
+
+import com.step902020.capstone.security.CurrentUser;
+import java.io.IOException;
+import java.util.*;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.util.Pair;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.servlet.view.RedirectView;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.function.Function;
+
+public class Recommender {
+
+  public static <E, U> List<E> recommend (U targetUser, List<U> users, Function<U, List<E>> getItemList, int numNeeded) {
+    Map<U, Integer> userToScore= new HashMap<U, Integer>();
+    Set<E> targetUserEvents = new HashSet(getItemList.apply(targetUser));
+    // for each user, calculate the total number of differences between the current user and the target user
+    // store this result as the distance
+    for (U user: users) {
+      if (user.equals(targetUser)) {
+        continue;
+      }
+      int dist = 0;
+      Set<E> userEvents = new HashSet(getItemList.apply(user));
+      for (E e : targetUserEvents) {
+        if (!userEvents.contains(e)) {
+          dist++;
+        }
+      }
+      for (E e : userEvents) {
+        if (!targetUserEvents.contains(e)) {
+          dist++;
+        }
+      }
+      userToScore.put(user, dist);
+    }
+    // sort the resulting list of individuals by increasing distance
+    // Create a list from elements of HashMap
+    List<Map.Entry<U, Integer>> sortedUsers =
```

Comment:
You might be interested in a `TreeMap` with a custom comparator. It's a sorted map; normally sorted by key but can also take a custom comparator, such as the one below.

Reply:
Correct me if I'm wrong, but I think TreeMap only keeps the keys in a sorted order, not values.

Reply:
By default, yes. But you can convince it otherwise with the same comparator you had before.

Reply:
Here is an example:
```
  @Test
  public void treeMapSortedByValue() {
    Map<String, Integer> unsortedMap = new HashMap<>();
    unsortedMap.put("z", 6);
    unsortedMap.put("b", 4);
    unsortedMap.put("t", 7);
    unsortedMap.put("e", 2);
    unsortedMap.put("a", 9);
    unsortedMap.put("g", 1);
    unsortedMap.put("w", 5);


    Map<String, Integer> treeMapDefaultSort = new TreeMap<>(unsortedMap);
    Comparator<String> valueComparator = (s1, s2) -> unsortedMap.get(s1) - unsortedMap.get(s2);
    Map<String, Integer> treeMapValueSort = new TreeMap<String, Integer>(valueComparator);
    treeMapValueSort.putAll(unsortedMap);

    // "a", "b", "e", "g", "t", "w", "z"
    assertArrayEquals(new Integer[] {9, 4, 2, 1, 7, 5, 6},
      treeMapDefaultSort.values().toArray(new Integer[0]));

    assertArrayEquals(new Integer[] {1, 2, 4, 5, 6, 7, 9},
        treeMapValueSort.values().toArray(new Integer[0]));

  }
```